src/v/cluster/shard_table.h

```diff
@@ -23,37 +23,81 @@ namespace cluster {
 /// \brief this is populated by consensus::controller
 /// every core will have a _full_ copy of all indexes
 class shard_table final {
+    struct shard_revision {
+        ss::shard_id shard;
+        model::revision_id revision;
+    };
+
 public:
     bool contains(const raft::group_id& group) {
         return _group_idx.find(group) != _group_idx.end();
     }
     ss::shard_id shard_for(const raft::group_id& group) {
-        return _group_idx.find(group)->second;
+        return _group_idx.find(group)->second.shard;
     }
 
     /**
      * \brief Lookup the owning shard for an ntp.
      */
     std::optional<ss::shard_id> shard_for(const model::ntp& ntp) {
         if (auto it = _ntp_idx.find(ntp); it != _ntp_idx.end()) {
-            return it->second;
+            return it->second.shard;
         }
         return std::nullopt;
     }
-    void insert(model::ntp ntp, ss::shard_id i) {
-        _ntp_idx.insert({std::move(ntp), i});
+
+    bool insert(model::ntp ntp, ss::shard_id i, model::revision_id rev) {
+        auto [_, success] = _ntp_idx.insert(
+          {std::move(ntp), shard_revision{i, rev}});
+        return success;
+    }
+
+    bool insert(raft::group_id g, ss::shard_id i, model::revision_id rev) {
+        auto [_, success] = _group_idx.insert({g, shard_revision{i, rev}});
+        return success;
+    }
+
+    void update(
+      const model::ntp& ntp,
+      raft::group_id g,
+      ss::shard_id shard,
+      model::revision_id rev) {
+        if (auto it = _ntp_idx.find(ntp); it != _ntp_idx.end()) {
+            if (it->second.revision > rev) {
+                return;
+            }
+        }
+        if (auto it = _group_idx.find(g); it != _group_idx.end()) {
+            if (it->second.revision > rev) {
+                return;
+            }
+        }
+
+        _ntp_idx.insert_or_assign(ntp, shard_revision{shard, rev});
+        _group_idx.insert_or_assign(g, shard_revision{shard, rev});
     }
-    void insert(raft::group_id g, ss::shard_id i) { _group_idx.insert({g, i}); }
 
-    void erase(const model::ntp& ntp, raft::group_id g) {
+    void
+    erase(const model::ntp& ntp, raft::group_id g, model::revision_id rev) {
+        if (auto it = _ntp_idx.find(ntp); it != _ntp_idx.end()) {
+            if (it->second.revision > rev) {
+                return;
+            }
+        }
+        if (auto it = _group_idx.find(g); it != _group_idx.end()) {
+            if (it->second.revision > rev) {
+                return;
+            }
+        }
+
         _ntp_idx.erase(ntp);
         _group_idx.erase(g);
     }
 
 private:
     // kafka index
-    absl::flat_hash_map<model::ntp, ss::shard_id> _ntp_idx;
+    absl::flat_hash_map<model::ntp, shard_revision> _ntp_idx;
```

Comment:
I think these maps deserve a comment about the eventually consistent data structure. 

Also, I think we should move them to `node_hash_map` since the keys are fairly big.

Reply:
good point